src/IO/S3/Client.cpp

```diff
@@ -703,19 +703,34 @@ Client::doRequest(RequestType & request, RequestFn request_fn) const
             continue;
         }
 
-        if (error.GetResponseCode() != Aws::Http::HttpResponseCode::MOVED_PERMANENTLY)
+        /// IllegalLocationConstraintException may indicate that we are working with an opt-in region (e.g. me-south-1)
+        /// In that case, we need to update the region and try again
+        if (error.GetResponseCode() != Aws::Http::HttpResponseCode::MOVED_PERMANENTLY && error.GetExceptionName() != "IllegalLocationConstraintException")
             return result;
 
         // maybe we detect a correct region
-        if (!detect_region)
+        bool new_region_detected = false;
+        if (!detect_region || error.GetExceptionName() == "IllegalLocationConstraintException")
```

Comment:
Nitpick and not required: Consider using a variable for this check `error.GetExceptionName() == "IllegalLocationConstraintException`. You re-wrote this in three differente places.

bool is_illegal_constraint_exception = error.GetExceptionName() == "IllegalLocationConstraintException;

if1 (... && !is_illegal

if2 (... && is_illegal

if3 (... && is_illegal